File: sdks/swift/Sources/BCS/Deserializer.swift
// Copyright (c) BCS SDK Contributors
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// BCS Deserializer - Manual deserialization API
public final class BcsDeserializer {
    @usableFromInline
    internal let data: [UInt8]
    @usableFromInline
    internal var offset: Int = 0
    @usableFromInline
    internal var depth: Int = 0

    /// Create a deserializer from a byte array
    @inlinable
    public init(_ data: [UInt8]) {
        self.data = data
    }

    /// Create a deserializer from Data
    @inlinable
    public init(_ data: Data) {
        self.data = Array(data)
    }

    // MARK: - Boolean

    /// Read a boolean value
    @inlinable
    public func readBool() throws -> Bool {
        let byte = try readU8()
        switch byte {
        case 0: return false
        case 1: return true
        default: throw BcsError.invalidBoolean(byte)
        }
    }

    // MARK: - Unsigned Integers

    /// Read an unsigned 8-bit integer
    @inlinable
    public func readU8() throws -> UInt8 {
        guard offset < data.count else {
            throw BcsError.unexpectedEof()
        }
        let value = data[offset]
        offset &+= 1
        return value
    }

    /// Read an unsigned 16-bit integer (little-endian)
    @inlinable
    public func readU16() throws -> UInt16 {
        guard offset &+ 2 <= data.count else {
            throw BcsError.unexpectedEof()
        }
        let value = UInt16(data[offset]) | (UInt16(data[offset &+ 1]) &<< 8)
        offset &+= 2
        return value
    }

    /// Read an unsigned 32-bit integer (little-endian)
    @inlinable
    public func readU32() throws -> UInt32 {
        guard offset &+ 4 <= data.count else {
            throw BcsError.unexpectedEof()
        }
        let value =
            UInt32(data[offset])
            | (UInt32(data[offset &+ 1]) &<< 8)
            | (UInt32(data[offset &+ 2]) &<< 16)
            | (UInt32(data[offset &+ 3]) &<< 24)
        offset &+= 4
        return value
    }

    /// Read an unsigned 64-bit integer (little-endian)
    @inlinable
    public func readU64() throws -> UInt64 {
        guard offset &+ 8 <= data.count else {
            throw BcsError.unexpectedEof()
        }
        let value =
            UInt64(data[offset])
            | (UInt64(data[offset &+ 1]) &<< 8)
            | (UInt64(data[offset &+ 2]) &<< 16)
            | (UInt64(data[offset &+ 3]) &<< 24)
            | (UInt64(data[offset &+ 4]) &<< 32)
            | (UInt64(data[offset &+ 5]) &<< 40)
            | (UInt64(data[offset &+ 6]) &<< 48)
            | (UInt64(data[offset &+ 7]) &<< 56)
        offset &+= 8
        return value
    }

    /// Read an unsigned 128-bit integer (little-endian byte array)
    @inlinable
    public func readU128() throws -> [UInt8] {
        try readFixedBytes(16)
    }

    /// Read an unsigned 256-bit integer (little-endian byte array)
    @inlinable
    public func readU256() throws -> [UInt8] {
        try readFixedBytes(32)
    }

    // MARK: - Signed Integers

    /// Read a signed 8-bit integer
    @inlinable
    public func readI8() throws -> Int8 {
        Int8(bitPattern: try readU8())
    }

    /// Read a signed 16-bit integer (little-endian)
    @inlinable
    public func readI16() throws -> Int16 {
        Int16(bitPattern: try readU16())
    }

    /// Read a signed 32-bit integer (little-endian)
    @inlinable
    public func readI32() throws -> Int32 {
        Int32(bitPattern: try readU32())
    }

    /// Read a signed 64-bit integer (little-endian)
    @inlinable
    public func readI64() throws -> Int64 {
        Int64(bitPattern: try readU64())
    }

    /// Read a signed 128-bit integer (little-endian byte array)
    @inlinable
    public func readI128() throws -> [UInt8] {
        try readU128()
    }

    /// Read a signed 256-bit integer (little-endian byte array)
    @inlinable
    public func readI256() throws -> [UInt8] {
        try readU256()
    }

    // MARK: - ULEB128

    /// Read a ULEB128-encoded value
    @inlinable
    public func readUleb128() throws -> UInt32 {
        guard offset < data.count else {
            throw BcsError.unexpectedEof()
        }

        // Fast path for single-byte values (0-127)
        let first = data[offset]
        if first < 0x80 {
            offset &+= 1
            return UInt32(first)
        }

        // Multi-byte path
        var value: UInt32 = UInt32(first & 0x7F)
        var shift: UInt32 = 7
        offset &+= 1

        for i in 1..<5 {
            guard offset < data.count else {
                throw BcsError.unexpectedEof()
            }

            let byte = data[offset]
            offset &+= 1
            let digit = byte & 0x7F

            value |= UInt32(digit) &<< shift

            // Check if this is the last byte (high bit not set)
            if (byte & 0x80) == 0 {
                // Check for non-canonical encoding (trailing zeros)
                if digit == 0 {
                    throw BcsError.nonCanonicalUleb128()
                }

                // Check for overflow on final byte
                if i == 4 && digit > 0x0F {
                    throw BcsError.uleb128Overflow()
                }

                return value
            }

            shift &+= 7
        }

        // If we've read 5 bytes and still have continuation bit, overflow
        throw BcsError.uleb128Overflow()
    }

    // MARK: - Bytes and Strings

    /// Read fixed-length bytes (without length prefix)
    @inlinable
    public func readFixedBytes(_ length: Int) throws -> [UInt8] {
        guard offset &+ length <= data.count else {
            throw BcsError.unexpectedEof()
        }
        let result = Array(data[offset..<(offset &+ length)])
        offset &+= length
        return result
    }

    /// Read bytes with ULEB128 length prefix
    @inlinable
    public func readBytes() throws -> [UInt8] {
        let length = try uleb128ToInt(try readUleb128())
        try checkSequenceLength(length)
        return try readFixedBytes(length)
    }

    /// Read a UTF-8 string with ULEB128 length prefix
    @inlinable
    public func readString() throws -> String {
        let length = try uleb128ToInt(try readUleb128())
        try checkSequenceLength(length)
        guard offset &+ length <= data.count else {
            throw BcsError.unexpectedEof()
        }
        // Create string directly from slice to avoid intermediate array allocation
        let endOffset = offset &+ length
        let str = data[offset..<endOffset].withUnsafeBufferPointer { buffer in
            String(decoding: buffer, as: UTF8.self)
        }
        // Validate UTF-8 by checking if round-trip produces same length
        if str.utf8.count != length {
            throw BcsError.invalidUtf8()
        }
        offset = endOffset
        return str
    }

    // MARK: - Composite Types

    /// Read an optional value
    @inlinable
    public func readOption<T>(_ deserializer: (BcsDeserializer) throws -> T) throws -> T? {
        let tag = try readU8()
        switch tag {
        case 0: return nil
        case 1: return try deserializer(self)
        default: throw BcsError.invalidOption(tag)
        }
    }

    /// Read a vector with element deserializer
    @inlinable
    public func readVector<T>(_ deserializer: (BcsDeserializer) throws -> T) throws -> [T] {
        let length = try uleb128ToInt(try readUleb128())
        try checkSequenceLength(length)

        var result: [T] = []
        result.reserveCapacity(length)
        for _ in 0..<length {
            result.append(try deserializer(self))
        }
        return result
    }

    /// Read a map with key/value deserializers
    public func readMap<K: Comparable, V>(
        keyDeserializer: (BcsDeserializer) throws -> K,
        valueDeserializer: (BcsDeserializer) throws -> V
    ) throws -> [K: V] {
        let length = try uleb128ToInt(try readUleb128())
        try checkSequenceLength(length)

        var result: [K: V] = [:]
        result.reserveCapacity(length)
        var prevKeyBytes: ArraySlice<UInt8>?

        for _ in 0..<length {
            // Remember position before reading key
            let keyStart = offset

            let key = try keyDeserializer(self)

            // Get key bytes for ordering check (use slice to avoid allocation)
            let keyBytes = data[keyStart..<offset]

            // Check ordering
            if let prev = prevKeyBytes {
                if keyBytes.lexicographicallyPrecedes(prev) || keyBytes.elementsEqual(prev) {
                    if keyBytes.elementsEqual(prev) {
                        throw BcsError.duplicateMapKey()
                    }
                    throw BcsError.nonCanonicalMap()
                }
            }
            prevKeyBytes = keyBytes

            let value = try valueDeserializer(self)
            result[key] = value
        }

        return result
    }

    /// Read an enum variant index
    @inlinable
    public func readVariantIndex() throws -> UInt32 {
        try readUleb128()
    }

    // MARK: - Container Depth

    /// Enter a struct/enum container (for depth tracking)
    @inlinable
    @discardableResult
    public func enterContainer() throws -> BcsDeserializer {
        if depth >= BcsConstants.maxContainerDepth {
            throw BcsError.exceededContainerDepth(depth + 1)
        }
        depth &+= 1
        return self
    }

    /// Leave a struct/enum container
    @inlinable
    @discardableResult
    public func leaveContainer() -> BcsDeserializer {
        if depth > 0 {
            depth -= 1
        }
        return self
    }

    // MARK: - State

    /// Check that all input has been consumed
    @inlinable
    public func checkEnd() throws {
        if offset < data.count {
            throw BcsError.remainingInput(data.count - offset)
        }
    }

    /// Get the current offset
    @inlinable
    public var currentOffset: Int {
        offset
    }

    /// Get the remaining bytes count
    @inlinable
    public var remaining: Int {
        data.count - offset
    }

    /// Check if there's more data to read
    @inlinable
    public var hasRemaining: Bool {
        offset < data.count
    }

    /// Peek at the next byte without consuming it
    @inlinable
    public func peek() throws -> UInt8 {
        guard offset < data.count else {
            throw BcsError.unexpectedEof()
        }
        return data[offset]
    }

    /// Skip n bytes
    @inlinable
    @discardableResult
    public func skip(_ n: Int) throws -> BcsDeserializer {
        guard offset &+ n <= data.count else {
            throw BcsError.unexpectedEof()
        }
        offset &+= n
        return self
    }

    /// Read fixed-length bytes as a slice (zero-copy view)
    @inlinable
    public func readFixedBytesSlice(_ length: Int) throws -> ArraySlice<UInt8> {
        guard offset &+ length <= data.count else {
            throw BcsError.unexpectedEof()
        }
        let result = data[offset..<(offset &+ length)]
        offset &+= length
        return result
    }

    // MARK: - Batch Read Operations

    /// Read a vector of U8 values efficiently
    @inlinable
    public func readU8Vector() throws -> [UInt8] {
        let length = try uleb128ToInt(try readUleb128())
        try checkSequenceLength(length)
        return try readFixedBytes(length)
    }

    /// Read a vector of U16 values efficiently
    @inlinable
    public func readU16Vector() throws -> [UInt16] {
        let uleb = try readUleb128()
        // Check for overflow: length * 2 must fit in Int
        guard uleb <= UInt32(Int.max / 2) else {
            throw BcsError.exceededMaxLength(Int(truncatingIfNeeded: uleb))
        }
        let length = Int(uleb)
        try checkSequenceLength(length)
        let byteCount = length * 2
        // Check bounds without overflow: ensure remaining bytes >= byteCount
        guard data.count - offset >= byteCount else {
            throw BcsError.unexpectedEof()
        }
        var result = [UInt16]()
        result.reserveCapacity(length)
        for _ in 0..<length {
            let value = UInt16(data[offset]) | (UInt16(data[offset &+ 1]) &<< 8)
            offset &+= 2
            result.append(value)
        }
        return result
    }

    /// Read a vector of U32 values efficiently
    @inlinable
    public func readU32Vector() throws -> [UInt32] {
        let uleb = try readUleb128()
        // Check for overflow: length * 4 must fit in Int
        guard uleb <= UInt32(Int.max / 4) else {
            throw BcsError.exceededMaxLength(Int(truncatingIfNeeded: uleb))
        }
        let length = Int(uleb)
        try checkSequenceLength(length)
        let byteCount = length * 4
        // Check bounds without overflow: ensure remaining bytes >= byteCount
        guard data.count - offset >= byteCount else {
            throw BcsError.unexpectedEof()
        }
        var result = [UInt32]()
        result.reserveCapacity(length)
        for _ in 0..<length {
            let value =
                UInt32(data[offset])
                | (UInt32(data[offset &+ 1]) &<< 8)
                | (UInt32(data[offset &+ 2]) &<< 16)
                | (UInt32(data[offset &+ 3]) &<< 24)
            offset &+= 4
            result.append(value)
        }
        return result
    }

    /// Read a vector of U64 values efficiently
    @inlinable
    public func readU64Vector() throws -> [UInt64] {
        let uleb = try readUleb128()
        // Check for overflow: length * 8 must fit in Int
        guard uleb <= UInt32(Int.max / 8) else {
            throw BcsError.exceededMaxLength(Int(truncatingIfNeeded: uleb))
        }
        let length = Int(uleb)
        try checkSequenceLength(length)
        let byteCount = length * 8
        // Check bounds without overflow: ensure remaining bytes >= byteCount
        guard data.count - offset >= byteCount else {
            throw BcsError.unexpectedEof()
        }
        var result = [UInt64]()
        result.reserveCapacity(length)
        for _ in 0..<length {
            let value =
                UInt64(data[offset])
                | (UInt64(data[offset &+ 1]) &<< 8)
                | (UInt64(data[offset &+ 2]) &<< 16)
                | (UInt64(data[offset &+ 3]) &<< 24)
                | (UInt64(data[offset &+ 4]) &<< 32)
                | (UInt64(data[offset &+ 5]) &<< 40)
                | (UInt64(data[offset &+ 6]) &<< 48)
                | (UInt64(data[offset &+ 7]) &<< 56)
            offset &+= 8
            result.append(value)
        }
        return result
    }

    // MARK: - Private

    /// Safely convert ULEB128 value to Int, handling 32-bit platform overflow
    @usableFromInline
    internal func uleb128ToInt(_ value: UInt32) throws -> Int {
        // On 32-bit platforms, Int.max is 2^31-1, so UInt32 values > Int.max would overflow
        guard value <= UInt32(Int.max) else {
            throw BcsError.exceededMaxLength(Int(truncatingIfNeeded: value))
        }
        return Int(value)
    }

    @usableFromInline
    internal func checkSequenceLength(_ length: Int) throws {
        if length > BcsConstants.maxSequenceLength {
            throw BcsError.exceededMaxLength(length)
        }
    }
}

File: sdks/go/bcs/deserializer.go
package bcs

import (
	"encoding/binary"
	"math/big"
	"sync"
	"unicode/utf8"
)

// Pre-computed constants for signed integer conversion to avoid repeated allocations.
var (
	signBit128  *big.Int
	signBit256  *big.Int
	modulus128  *big.Int
	modulus256  *big.Int
	initBigOnce sync.Once
)

func initBigConstants() {
	initBigOnce.Do(func() {
		signBit128 = new(big.Int).Lsh(big.NewInt(1), 127)
		signBit256 = new(big.Int).Lsh(big.NewInt(1), 255)
		modulus128 = new(big.Int).Lsh(big.NewInt(1), 128)
		modulus256 = new(big.Int).Lsh(big.NewInt(1), 256)
	})
}

// Deserializer provides methods for BCS deserialization.
type Deserializer struct {
	data   []byte
	offset int
	depth  int
}

// deserializerPool provides reusable Deserializer instances.
var deserializerPool = sync.Pool{
	New: func() interface{} {
		return &Deserializer{}
	},
}

// NewDeserializer creates a new BCS deserializer.
func NewDeserializer(data []byte) *Deserializer {
	initBigConstants()
	return &Deserializer{data: data, offset: 0}
}

// AcquireDeserializer gets a Deserializer from the pool and initializes it with data.
// Call ReleaseDeserializer when done to return it to the pool.
func AcquireDeserializer(data []byte) *Deserializer {
	initBigConstants()
	d := deserializerPool.Get().(*Deserializer)
	d.data = data
	d.offset = 0
	return d
}

// ReleaseDeserializer returns a Deserializer to the pool for reuse.
func ReleaseDeserializer(d *Deserializer) {
	d.data = nil
	d.offset = 0
	deserializerPool.Put(d)
}

// ==========================================================================
// BOOLEAN
// ==========================================================================

// ReadBool deserializes a boolean value.
func (d *Deserializer) ReadBool() (bool, error) {
	if err := d.ensureBytes(1); err != nil {
		return false, err
	}
	b := d.data[d.offset]
	d.offset++
	switch b {
	case 0:
		return false, nil
	case 1:
		return true, nil
	default:
		return false, NewInvalidBoolean(b)
	}
}

// ==========================================================================
// UNSIGNED INTEGERS
// ==========================================================================

// ReadU8 deserializes an unsigned 8-bit integer.
func (d *Deserializer) ReadU8() (uint8, error) {
	if err := d.ensureBytes(1); err != nil {
		return 0, err
	}
	value := d.data[d.offset]
	d.offset++
	return value, nil
}

// ReadU16 deserializes an unsigned 16-bit integer (little-endian).
func (d *Deserializer) ReadU16() (uint16, error) {
	if err := d.ensureBytes(2); err != nil {
		return 0, err
	}
	value := binary.LittleEndian.Uint16(d.data[d.offset:])
	d.offset += 2
	return value, nil
}

// ReadU32 deserializes an unsigned 32-bit integer (little-endian).
func (d *Deserializer) ReadU32() (uint32, error) {
	if err := d.ensureBytes(4); err != nil {
		return 0, err
	}
	value := binary.LittleEndian.Uint32(d.data[d.offset:])
	d.offset += 4
	return value, nil
}

// ReadU64 deserializes an unsigned 64-bit integer (little-endian).
func (d *Deserializer) ReadU64() (uint64, error) {
	if err := d.ensureBytes(8); err != nil {
		return 0, err
	}
	value := binary.LittleEndian.Uint64(d.data[d.offset:])
	d.offset += 8
	return value, nil
}

// ReadU128 deserializes an unsigned 128-bit integer (little-endian).
func (d *Deserializer) ReadU128() (*big.Int, error) {
	if err := d.ensureBytes(16); err != nil {
		return nil, err
	}
	return d.readBigIntLE(16), nil
}

// ReadU256 deserializes an unsigned 256-bit integer (little-endian).
func (d *Deserializer) ReadU256() (*big.Int, error) {
	if err := d.ensureBytes(32); err != nil {
		return nil, err
	}
	return d.readBigIntLE(32), nil
}

// ==========================================================================
// SIGNED INTEGERS
// ==========================================================================

// ReadI8 deserializes a signed 8-bit integer (two's complement).
func (d *Deserializer) ReadI8() (int8, error) {
	u, err := d.ReadU8()
	if err != nil {
		return 0, err
	}
	return int8(u), nil
}

// ReadI16 deserializes a signed 16-bit integer (two's complement, little-endian).
func (d *Deserializer) ReadI16() (int16, error) {
	u, err := d.ReadU16()
	if err != nil {
		return 0, err
	}
	return int16(u), nil
}

// ReadI32 deserializes a signed 32-bit integer (two's complement, little-endian).
func (d *Deserializer) ReadI32() (int32, error) {
	u, err := d.ReadU32()
	if err != nil {
		return 0, err
	}
	return int32(u), nil
}

// ReadI64 deserializes a signed 64-bit integer (two's complement, little-endian).
func (d *Deserializer) ReadI64() (int64, error) {
	u, err := d.ReadU64()
	if err != nil {
		return 0, err
	}
	return int64(u), nil
}

// ReadI128 deserializes a signed 128-bit integer (two's complement, little-endian).
// Uses cached constants to avoid repeated allocations.
func (d *Deserializer) ReadI128() (*big.Int, error) {
	unsigned, err := d.ReadU128()
	if err != nil {
		return nil, err
	}
	if unsigned.Cmp(signBit128) >= 0 {
		return new(big.Int).Sub(unsigned, modulus128), nil
	}
	return unsigned, nil
}

// ReadI256 deserializes a signed 256-bit integer (two's complement, little-endian).
// Uses cached constants to avoid repeated allocations.
func (d *Deserializer) ReadI256() (*big.Int, error) {
	unsigned, err := d.ReadU256()
	if err != nil {
		return nil, err
	}
	if unsigned.Cmp(signBit256) >= 0 {
		return new(big.Int).Sub(unsigned, modulus256), nil
	}
	return unsigned, nil
}

// ==========================================================================
// ULEB128
// ==========================================================================

// ReadULEB128 deserializes a ULEB128-encoded unsigned integer.
// Inlined for performance on the hot path.
func (d *Deserializer) ReadULEB128() (uint32, error) {
	// Fast path: single byte (values 0-127)
	if d.offset >= len(d.data) {
		return 0, NewUnexpectedEOF()
	}
	b := d.data[d.offset]
	if b < 0x80 {
		d.offset++
		return uint32(b), nil
	}

	// Slow path: multi-byte encoding
	value, bytesRead, err := DecodeULEB128WithOffset(d.data, d.offset)
	if err != nil {
		return 0, err
	}
	d.offset += bytesRead
	return value, nil
}

// ==========================================================================
// BYTES AND STRINGS
// ==========================================================================

// ReadBytes deserializes a byte slice (length-prefixed with ULEB128).
func (d *Deserializer) ReadBytes() ([]byte, error) {
	length, err := d.ReadULEB128()
	if err != nil {
		return nil, err
	}
	if length > MaxSequenceLength {
		return nil, NewExceededMaxLength(uint64(length))
	}
	return d.ReadFixedBytes(int(length))
}

// ReadString deserializes a UTF-8 string (length-prefixed with ULEB128).
func (d *Deserializer) ReadString() (string, error) {
	b, err := d.ReadBytes()
	if err != nil {
		return "", err
	}
	if !utf8.Valid(b) {
		return "", NewInvalidUTF8("")
	}
	return string(b), nil
}

// ReadFixedBytes deserializes fixed-length bytes (no length prefix).
// Returns a copy of the bytes.
func (d *Deserializer) ReadFixedBytes(length int) ([]byte, error) {
	if err := d.ensureBytes(length); err != nil {
		return nil, err
	}
	result := make([]byte, length)
	copy(result, d.data[d.offset:d.offset+length])
	d.offset += length
	return result, nil
}

// ReadFixedBytesNoCopy deserializes fixed-length bytes without copying.
// WARNING: The returned slice is a view into the deserializer's internal buffer.
// Do not modify the returned slice and do not use it after the deserializer
// is released or reused. Use ReadFixedBytes if you need to own the data.
func (d *Deserializer) ReadFixedBytesNoCopy(length int) ([]byte, error) {
	if err := d.ensureBytes(length); err != nil {
		return nil, err
	}
	result := d.data[d.offset : d.offset+length]
	d.offset += length
	return result, nil
}

// ==========================================================================
// OPTION
// ==========================================================================

// ReadOptionTag reads the option tag and returns whether a value is present.
func (d *Deserializer) ReadOptionTag() (bool, error) {
	if err := d.ensureBytes(1); err != nil {
		return false, err
	}
	tag := d.data[d.offset]
	d.offset++
	switch tag {
	case 0:
		return false, nil
	case 1:
		return true, nil
	default:
		return false, NewInvalidOption(tag)
	}
}

// ==========================================================================
// VECTOR
// ==========================================================================

// ReadVectorLen reads the length prefix for a vector.
func (d *Deserializer) ReadVectorLen() (uint32, error) {
	length, err := d.ReadULEB128()
	if err != nil {
		return 0, err
	}
	if length > MaxSequenceLength {
		return 0, NewExceededMaxLength(uint64(length))
	}
	return length, nil
}

// ==========================================================================
// ENUM
// ==========================================================================

// ReadVariantIndex reads an enum variant index (ULEB128).
func (d *Deserializer) ReadVariantIndex() (uint32, error) {
	return d.ReadULEB128()
}

// ==========================================================================
// CONTAINER DEPTH
// ==========================================================================

// ContainerDepth returns the current container nesting depth.
func (d *Deserializer) ContainerDepth() int {
	return d.depth
}

// EnterStruct enters a struct container for depth tracking.
// Returns an error if container depth exceeds MaxContainerDepth (500).
func (d *Deserializer) EnterStruct() error {
	return d.enterContainer("struct")
}

// LeaveStruct leaves a struct container.
func (d *Deserializer) LeaveStruct() {
	d.leaveContainer()
}

// EnterEnum enters an enum container for depth tracking and reads the variant index.
// Returns the variant index and an error if container depth exceeds MaxContainerDepth (500).
func (d *Deserializer) EnterEnum() (uint32, error) {
	if err := d.enterContainer("enum"); err != nil {
		return 0, err
	}
	return d.ReadVariantIndex()
}

// LeaveEnum leaves an enum container.
func (d *Deserializer) LeaveEnum() {
	d.leaveContainer()
}

func (d *Deserializer) enterContainer(containerType string) error {
	if d.depth >= MaxContainerDepth {
		return NewExceededContainerDepth(containerType)
	}
	d.depth++
	return nil
}

func (d *Deserializer) leaveContainer() {
	if d.depth > 0 {
		d.depth--
	}
}

// ==========================================================================
// MAP
// ==========================================================================

// ReadMapLen reads the length prefix for a map.
func (d *Deserializer) ReadMapLen() (uint32, error) {
	return d.ReadVectorLen()
}

// Position returns the current read position for key comparison.
func (d *Deserializer) Position() int {
	return d.offset
}

// SliceFrom returns a slice of the data from start to current position.
func (d *Deserializer) SliceFrom(start int) []byte {
	return d.data[start:d.offset]
}

// ==========================================================================
// BATCH OPERATIONS
// ==========================================================================

// ReadU8Slice reads a vector of u8 values (length-prefixed).
// More efficient than calling ReadVectorLen + ReadU8 in a loop.
// Returns a copy of the data.
func (d *Deserializer) ReadU8Slice() ([]byte, error) {
	length, err := d.ReadVectorLen()
	if err != nil {
		return nil, err
	}
	return d.ReadFixedBytes(int(length))
}

// ReadU8SliceNoCopy reads a vector of u8 values without copying.
// WARNING: The returned slice is a view into the deserializer's internal buffer.
func (d *Deserializer) ReadU8SliceNoCopy() ([]byte, error) {
	length, err := d.ReadVectorLen()
	if err != nil {
		return nil, err
	}
	return d.ReadFixedBytesNoCopy(int(length))
}

// ReadU16Slice reads a vector of u16 values (length-prefixed).
func (d *Deserializer) ReadU16Slice() ([]uint16, error) {
	length, err := d.ReadVectorLen()
	if err != nil {
		return nil, err
	}
	if err := d.ensureBytes(int(length) * 2); err != nil {
		return nil, err
	}
	result := make([]uint16, length)
	for i := range result {
		result[i] = binary.LittleEndian.Uint16(d.data[d.offset:])
		d.offset += 2
	}
	return result, nil
}

// ReadU32Slice reads a vector of u32 values (length-prefixed).
func (d *Deserializer) ReadU32Slice() ([]uint32, error) {
	length, err := d.ReadVectorLen()
	if err != nil {
		return nil, err
	}
	if err := d.ensureBytes(int(length) * 4); err != nil {
		return nil, err
	}
	result := make([]uint32, length)
	for i := range result {
		result[i] = binary.LittleEndian.Uint32(d.data[d.offset:])
		d.offset += 4
	}
	return result, nil
}

// ReadU64Slice reads a vector of u64 values (length-prefixed).
func (d *Deserializer) ReadU64Slice() ([]uint64, error) {
	length, err := d.ReadVectorLen()
	if err != nil {
		return nil, err
	}
	if err := d.ensureBytes(int(length) * 8); err != nil {
		return nil, err
	}
	result := make([]uint64, length)
	for i := range result {
		result[i] = binary.LittleEndian.Uint64(d.data[d.offset:])
		d.offset += 8
	}
	return result, nil
}

// ReadStringSlice reads a vector of strings (length-prefixed).
func (d *Deserializer) ReadStringSlice() ([]string, error) {
	length, err := d.ReadVectorLen()
	if err != nil {
		return nil, err
	}
	result := make([]string, length)
	for i := range result {
		result[i], err = d.ReadString()
		if err != nil {
			return nil, err
		}
	}
	return result, nil
}

// ==========================================================================
// UTILITY
// ==========================================================================

// CheckEnd verifies that all input has been consumed.
func (d *Deserializer) CheckEnd() error {
	if d.offset < len(d.data) {
		return NewRemainingInput(len(d.data) - d.offset)
	}
	return nil
}

// Remaining returns the number of remaining bytes.
func (d *Deserializer) Remaining() int {
	return len(d.data) - d.offset
}

// ==========================================================================
// PRIVATE HELPERS
// ==========================================================================

func (d *Deserializer) ensureBytes(count int) error {
	if d.offset+count > len(d.data) {
		return NewUnexpectedEOFWithDetails(count, len(d.data)-d.offset)
	}
	return nil
}

func (d *Deserializer) readBigIntLE(byteLength int) *big.Int {
	// Use stack-allocated arrays for common sizes to avoid heap allocation
	switch byteLength {
	case 16:
		var b [16]byte
		for i := 0; i < 16; i++ {
			b[15-i] = d.data[d.offset+i]
		}
		d.offset += 16
		return new(big.Int).SetBytes(b[:])
	case 32:
		var b [32]byte
		for i := 0; i < 32; i++ {
			b[31-i] = d.data[d.offset+i]
		}
		d.offset += 32
		return new(big.Int).SetBytes(b[:])
	default:
		// Fallback for other sizes (rare)
		b := make([]byte, byteLength)
		for i := 0; i < byteLength; i++ {
			b[byteLength-1-i] = d.data[d.offset+i]
		}
		d.offset += byteLength
		return new(big.Int).SetBytes(b)
	}
}
